### PR TITLE
[TechDocs] Use DocsSyncroniser logger to get doc build logs

### LINKED ISSUE #10335: [TechDocs] Output logs from the sync process to stdout
## Feature Suggestion

The sync-process logs sent to the event stream as implemented in https://github.com/backstage/backstage/pull/6341 are really useful, and we would like to log them as part of running our backend. However, there are two problems with this- 
1. The logs go to the event stream and not stdout
2. The logs aren't formatted as logs

## Possible Implementation
Provide an additional logger to the sync endpoint

## Context

We would like to see these logs in the backend for troubleshooting purposes, to aid users without them having to save a screenshot of what errors they saw

### PR DESCRIPTION
**Pull request by bridgetbarnes**
## Hey, I just made a Pull Request!
Closes #10335

This adds the DocsSyncroniser logger as an additional transport to the logger that streams build events to the frontend, so they can also be used from the backend. 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by bridgetbarnes** [comment]
Hello! It seems like there might have been something not to do with these changes that caused the two action failures, e.g. Error: Input required and not supplied: github-token seems like it might be an intermittent error- could I have some assistance rerunning these please?

**Comment by iamEAP** [comment]
> _...seems like it might be an intermittent error- could I have some assistance rerunning these please?_

You got it!

> _I'm keen to have a go at this to get it sorted, but keen for some guidance before coming back with a PR that's not what the team is looking for_

Thanks for your patience on us getting back to you on this!  The approach in this PR seems reasonable, but my primary concern is that the overall backend logs could become quite verbose for those who are less interested in having these logs come through.

I see a couple of possible ways to take that into consideration:

**Configuration**
You could imagine having a new configuration setting like `techdocs.captureBuildLogs`, which you could use to determine whether or not to add the logger as you've done in this PR.

**Injection**
A more flexible option would be to introduce a new, optional `buildLogger` (or similarly named) argument to the `createRouter` function (perhaps only on the `OutOfTheBoxDeploymentOptions` type, since it's only relevant for "prepared/generated in backstage backend" deployment configuration).  This could then be passed into the `DocsSynchronizer` and used as you've done it here (and it could either default to a void logger, or only conditionally log if it's set).

I'd maybe encourage the latter, as it not only gives the flexibility of "do I want these logs or not," but also "can I treat these logs differently?"  You could achieve the same effect as you have here by passing the logger twice:

```ts
// Capture build logs in the same place / format as TechDocs backend logs
await createRouter({
  // ...
  logger: env.logger,
  buildLogger: env.logger,
});

// Capture build logs in a distinct way
await createRouter({
  // ...
  logger: env.logger,
  buildLogger: customLogger,
});

// Do not capture build logs (same behavior as today)
await createRouter({
  // ...
  logger: env.logger,
});
```

**Comment by bridgetbarnes** [comment]
Thank you so much! The injection method sounds really interesting, I'll take a stab at implementing it over the weekend

**Comment by bridgetbarnes** [comment]
Hello @emmaindal! I think I have the injection working, I've just pushed the changes (will fix up the signing now)

File: .changeset/quiet-paws-study.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
+---
+'@backstage/plugin-techdocs-backend': patch
+---
+
+Output logs from a TechDocs build to a logging transport in addition to existing frontend event stream, for capturing these logs to other sources.
+
+This allows users to capture debugging information around why tech docs fail to build without needing to rely on end users ending information from their web browser.
```
Comment (Niksko) [inline .changeset/quiet-paws-study.md:7]:
```suggestion
This allows users to capture debugging information around why tech docs fail to build without needing to rely on end users capturing information from their web browser.
```

File: .changeset/quiet-paws-study.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+The most common use case is to log to the same place as the rest of the backend application logs.
+
+To use, include
```
Comment (Niksko) [inline .changeset/quiet-paws-study.md:11]:
```suggestion
Sample usage:
```

File: .changeset/quiet-paws-study.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  Publisher,
+} from '@backstage/plugin-techdocs-backend';
+import Docker from 'dockerode';
+import { Router } from 'express';
+import { PluginEnvironment } from '../types';
+
+export default async function createPlugin(
+  env: PluginEnvironment,
+): Promise<Router> {
+  // Preparers are responsible for fetching source files for documentation.
```
Comment (Niksko) [inline .changeset/quiet-paws-study.md:28]: I think remove these comments, and just comment the line where the `buildTransportLogger` is. That makes it clear that that's the interesting part

File: .changeset/quiet-paws-study.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+  // checks if the publisher is working and logs the result
+  await publisher.getReadiness();
+
+  return await createRouter({
+    preparers,
+    generators,
+    publisher,
+    logger: env.logger,
+    buildTransportLogger: env.logger,
```
Comment (Niksko) [inline .changeset/quiet-paws-study.md:60]:
```suggestion
    buildLogTransport: env.logger,
```

**Review by iamEAP — APPROVED** [review]
Niiice! This is great. Thanks again for your contribution.

Over to you for a final look and merge, @camilaibs

**Review by camilaibs — APPROVED** [review]

File: .changeset/quiet-paws-study.md (hunk 3 above)
Comment (camilaibs) [reply .changeset/quiet-paws-study.md]: Comments removed, thanks again 🙂

**Pull request merged**